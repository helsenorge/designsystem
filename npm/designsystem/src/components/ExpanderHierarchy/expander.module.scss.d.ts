export type Styles = {
  expander: string;
  expander__button: string;
  'expander__button--2-and-lower': string;
  'expander__button--print': string;
  expander__content: string;
  'expander__content--1': string;
  'expander__content--2': string;
  'expander__content--3-and-lower': string;
  expander__icon: string;
  'expander__icon--2-and-lower': string;
  'expander__icon--3-and-lower': string;
  expander__title: string;
  'expander__title--1': string;
  'expander__title--2': string;
  'expander__title--3': string;
  'expander__title--4-and-lower': string;
  'expander__title--expanded': string;
  'expander__title--print': string;
  'fade-in': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
