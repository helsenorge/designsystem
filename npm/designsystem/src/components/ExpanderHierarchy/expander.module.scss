@use 'sass:map';
@import '../../scss/breakpoints';
@import '../../scss/palette';
@import '../../scss/spacers';
@import '../../scss/title';

.expander {
  $item: &;

  @media (prefers-reduced-motion: no-preference) {
    animation: fade-in 150ms linear forwards;
  }

  &:first-of-type {
    border-top: 1px solid $neutral500;
  }

  &__button {
    all: unset;
    box-sizing: border-box;
    display: flex;
    justify-content: space-between;
    align-items: center;
    width: 100%;
    cursor: pointer;
    background-color: transparent;
    border-bottom: 1px solid $neutral500;
    padding: getSpacer(xs) 0 getSpacer(xs) 0;

    @media (min-width: map.get($grid-breakpoints, lg)) {
      flex-direction: row-reverse;
      justify-content: start;
      padding: getSpacer(s) getSpacer(s) getSpacer(s) getSpacer(2xs);
    }

    &:hover,
    &:focus {
      background-color: $neutral100;
    }

    &:focus {
      text-decoration: underline;
      outline: getSpacer(4xs) solid $black;
    }

    &[aria-expanded='true'] {
      background-color: $neutral100;
      border-bottom: none;
    }

    @media print {
      &--print {
        background-color: $neutral100;
      }
    }

    &--2-and-lower {
      padding: getSpacer(2xs) 0 getSpacer(2xs) 0;

      @media (min-width: map.get($grid-breakpoints, lg)) {
        padding: getSpacer(xs) getSpacer(s) getSpacer(xs) getSpacer(2xs);
      }
    }
  }

  &__title {
    margin: 0;

    &--1 {
      @include title2;
    }

    &--2 {
      @include title3;
    }

    &--3 {
      @include title4;
    }

    &--4-and-lower {
      @include title5;
    }

    &--expanded {
      + #{$item}__content {
        display: block;
      }
    }

    @media print {
      &--print {
        + #{$item}__content {
          display: block;
        }
      }
    }
  }

  &__content {
    display: none;
    border-bottom: 1px solid $neutral500;
    padding-top: getSpacer(m);
    padding-left: getSpacer(2xs);
    padding-bottom: getSpacer(m);

    &--1 {
      padding-top: getSpacer(xl);
    }

    &--1,
    &--2 {
      @media (min-width: map.get($grid-breakpoints, lg)) {
        padding-left: getSpacer(2xl);
      }
    }

    &--3-and-lower {
      @media (min-width: map.get($grid-breakpoints, lg)) {
        padding-left: getSpacer(xl);
      }
    }
  }

  &__icon {
    margin-left: getSpacer(s);

    &--2-and-lower {
      margin-left: getSpacer(xs);
    }

    @media (min-width: map.get($grid-breakpoints, lg)) {
      margin-left: 0;
      margin-right: getSpacer(s);
    }

    &--3-and-lower {
      @media (min-width: map.get($grid-breakpoints, lg)) {
        margin-right: getSpacer(xs);
      }
    }
  }
}

@keyframes fade-in {
  0% {
    opacity: 0;
    max-height: 0;
  }

  99% {
    max-height: 24rem;
  }

  100% {
    opacity: 1;
    max-height: none;
  }
}
