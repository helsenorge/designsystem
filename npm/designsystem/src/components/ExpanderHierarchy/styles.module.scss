@import '../../scss/palette';
@import '../../scss/spacers';
@import '../../scss/font-mixins';

.expanderhierarchy {
  list-style: none;
  padding: 0;
  margin: 0;

  &--2 {
    margin-top: getSpacer(xl);
  }

  &--3-and-lower {
    margin-top: getSpacer(m);
  }
}
