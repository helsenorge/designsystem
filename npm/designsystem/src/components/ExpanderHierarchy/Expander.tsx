import React from 'react';

import classNames from 'classnames';

import ExpanderButton from './ExpanderButton';
import ExpanderHierarchy, { ExpanderHierarchyCommonProps, ExpanderHierarchyProps } from './ExpanderHierarchy';
import { getHeadingTag } from './utils';
import { isComponent } from '../../utils/component';

import styles from './expander.module.scss';

export interface ExpanderProps extends ExpanderHierarchyCommonProps {
  /** Sets the trigger title */
  title: string;
  /** Opens or closes the expander */
  expanded?: boolean;
  /** Called when expander is open/closed. */
  onExpand?: (isExpanded: boolean) => void;
  /** Sets the expanded content */
  children?: React.ReactNode;
  /** Sets the data-testid attribute on the expander button. */
  testId?: string;
  /** Adds custom classes to the element. */
  className?: string;
}

export type ExpanderType = React.FC<ExpanderProps>;

const Expander: ExpanderType = ({
  title,
  htmlMarkup = 'h2',
  level = 1,
  print,
  expanded = false,
  onExpand,
  children,
  testId,
  className,
}: ExpanderProps) => {
  const contentClasses = classNames(
    styles.expander__content,
    (level === 1 || level === 2) && styles[`expander__content--${level}`],
    level > 2 && styles[`expander__content--3-and-lower`]
  );

  return (
    <li className={classNames(styles.expander, className)}>
      <ExpanderButton htmlMarkup={htmlMarkup} level={level} print={print} expanded={expanded} onExpand={onExpand} testId={testId}>
        {title}
      </ExpanderButton>
      <div className={contentClasses}>
        {React.Children.map(children, child =>
          isComponent<ExpanderHierarchyProps>(child, ExpanderHierarchy)
            ? React.cloneElement(child, { htmlMarkup: getHeadingTag(htmlMarkup), level: level + 1, print })
            : child
        )}
      </div>
    </li>
  );
};

export default Expander;
