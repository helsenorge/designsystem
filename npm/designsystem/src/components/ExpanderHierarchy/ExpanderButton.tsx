import React from 'react';

import classNames from 'classnames';

import { AnalyticsId } from '../../constants';
import { useExpand } from '../../hooks/useExpand';
import { useHover } from '../../hooks/useHover';
import { ExpanderHierarchyCommonProps } from '../ExpanderHierarchy';
import Icon, { IconSize } from '../Icons';
import ChevronDown from '../Icons/ChevronDown';
import ChevronUp from '../Icons/ChevronUp';

import styles from './expander.module.scss';

interface ExpanderButtonProps extends ExpanderHierarchyCommonProps {
  expanded: boolean;
  onExpand?: (isExpanded: boolean) => void;
  children: string;
  testId?: string;
}

const ExpanderButton: React.FC<ExpanderButtonProps> = ({ htmlMarkup = 'h2', level = 1, print, expanded, onExpand, children, testId }) => {
  const { hoverRef, isHovered } = useHover<HTMLButtonElement>();
  const [isExpanded, setIsExpanded] = useExpand(expanded, onExpand);

  const handleClick = (): void => {
    setIsExpanded(!isExpanded);
  };

  const CustomTag = htmlMarkup;

  const buttonClasses = classNames(
    styles.expander__button,
    level > 1 && styles[`expander__button--2-and-lower`],
    print && styles['expander__button--print']
  );

  const titleClasses = classNames(
    styles.expander__title,
    level <= 3 && styles[`expander__title--${level}`],
    level > 3 && styles[`expander__title--4-and-lower`],
    isExpanded && styles['expander__title--expanded'],
    print && styles['expander__title--print']
  );

  const iconClasses = classNames(
    styles.expander__icon,
    level > 1 && styles[`expander__icon--2-and-lower`],
    level > 2 && styles[`expander__icon--3-and-lower`]
  );

  return (
    <CustomTag className={titleClasses}>
      <button
        type="button"
        onClick={handleClick}
        className={buttonClasses}
        aria-expanded={isExpanded}
        ref={hoverRef}
        data-testid={testId}
        data-analyticsid={AnalyticsId.ExpanderHierarchyExpander}
      >
        {children}
        <Icon svgIcon={isExpanded ? ChevronUp : ChevronDown} isHovered={isHovered} className={iconClasses} size={IconSize.XSmall} />
      </button>
    </CustomTag>
  );
};

export default ExpanderButton;
