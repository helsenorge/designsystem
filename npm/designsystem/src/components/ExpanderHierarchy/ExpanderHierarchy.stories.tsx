import React from 'react';

import { ComponentStory, ComponentMeta } from '@storybook/react';

import ExpanderHierarchy from './ExpanderHierarchy';
import { shortLoremText, mediumLoremText } from '../../utils/loremtext';
import GridExample from '../GridExample';

export default {
  title: '@helsenorge∕designsystem-react/Components/ExpanderHierarchy',
  component: ExpanderHierarchy,
  parameters: {
    docs: {
      description: {
        component:
          'ExpanderList allows the creation of a single list of items which can be used as triggers to expand areas within the list to display arbitrary content.',
      },
    },
  },
  argTypes: {
    title: {
      control: 'text',
      defaultValue: 'Nivå',
    },
    htmlMarkup: {
      control: 'select',
      options: ['h2', 'h3', 'h4', 'h5', 'h6'],
      defaultValue: 'h2',
    },
    expanded: {
      control: 'boolean',
      defaultValue: false,
    },
    print: {
      control: 'boolean',
      defaultValue: false,
    },
  },
} as ComponentMeta<typeof ExpanderHierarchy & typeof ExpanderHierarchy.Expander>;

export const Default: ComponentStory<typeof ExpanderHierarchy & typeof ExpanderHierarchy.Expander> = args => (
  <GridExample>
    <ExpanderHierarchy {...args}>
      <ExpanderHierarchy.Expander title={`${args.title} (1)`} expanded={args.expanded}>
        {shortLoremText}
        <ExpanderHierarchy>
          <ExpanderHierarchy.Expander title={`${args.title} (2)`} expanded={args.expanded}>
            {mediumLoremText}
            <ExpanderHierarchy>
              <ExpanderHierarchy.Expander title={`${args.title} (3)`} expanded={args.expanded}>
                {mediumLoremText}
                <ExpanderHierarchy>
                  <ExpanderHierarchy.Expander title={`${args.title} (4)`} expanded={args.expanded}>
                    {mediumLoremText}
                    <ExpanderHierarchy>
                      <ExpanderHierarchy.Expander title={`${args.title} (5)`} expanded={args.expanded}>
                        {mediumLoremText}
                        <ExpanderHierarchy>
                          <ExpanderHierarchy.Expander title={`${args.title} (6)`} expanded={args.expanded}>
                            {mediumLoremText}{' '}
                            <ExpanderHierarchy>
                              <ExpanderHierarchy.Expander title={`${args.title} (7)`} expanded={args.expanded}>
                                {mediumLoremText}
                              </ExpanderHierarchy.Expander>
                            </ExpanderHierarchy>
                          </ExpanderHierarchy.Expander>
                        </ExpanderHierarchy>
                      </ExpanderHierarchy.Expander>
                    </ExpanderHierarchy>
                  </ExpanderHierarchy.Expander>
                </ExpanderHierarchy>
              </ExpanderHierarchy.Expander>
            </ExpanderHierarchy>
          </ExpanderHierarchy.Expander>
          <ExpanderHierarchy.Expander title={`${args.title} (2)`} expanded={args.expanded}>
            {mediumLoremText}
          </ExpanderHierarchy.Expander>
          <ExpanderHierarchy.Expander title={`${args.title} (2)`} expanded={args.expanded}>
            {mediumLoremText}
          </ExpanderHierarchy.Expander>
          <ExpanderHierarchy.Expander title={`${args.title} (2)`} expanded={args.expanded}>
            {mediumLoremText}
          </ExpanderHierarchy.Expander>
        </ExpanderHierarchy>
      </ExpanderHierarchy.Expander>
    </ExpanderHierarchy>
  </GridExample>
);
