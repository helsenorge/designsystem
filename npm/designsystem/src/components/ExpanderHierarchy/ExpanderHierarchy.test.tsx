import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import ExpanderHierarchy from '../ExpanderHierarchy';

describe('ExpanderHierarchy', () => {
  describe('gitt at ExpanderHierarchy har en Expander', () => {
    describe('når komponenten rendres', () => {
      it('så ser den riktig ut', async () => {
        const { container } = render(
          <ExpanderHierarchy>
            <ExpanderHierarchy.Expander title="Nivå 1">{'hei'}</ExpanderHierarchy.Expander>
          </ExpanderHierarchy>
        );

        expect(container).toMatchSnapshot();
      });
    });

    describe('når komponenten rendres med print=true', () => {
      it('så ser den riktig ut', async () => {
        const { container } = render(
          <ExpanderHierarchy print>
            <ExpanderHierarchy.Expander title="Nivå 1">{'hei'}</ExpanderHierarchy.Expander>
          </ExpanderHierarchy>
        );

        expect(container).toMatchSnapshot();
      });
    });

    describe('når man åpner en Expander', () => {
      it('så vises innholdet i den', async () => {
        render(
          <ExpanderHierarchy>
            <ExpanderHierarchy.Expander title="Nivå 1">{'hei'}</ExpanderHierarchy.Expander>
          </ExpanderHierarchy>
        );

        const heading = screen.getByRole('heading', { name: 'Nivå 1', level: 2 });
        expect(heading).toBeVisible();
        const button = screen.getByRole('button', { name: 'Nivå 1' });
        expect(button).toBeVisible();

        expect(button).toHaveAttribute('aria-expanded', 'false');

        await userEvent.click(button);
        expect(button).toHaveAttribute('aria-expanded', 'true');

        const content = screen.getByText('hei');
        expect(content).toBeVisible();
      });
    });

    describe('når Expander har onExpand-callback', () => {
      it('så vises innholdet i den', async () => {
        const handleExpand = vi.fn();

        render(
          <ExpanderHierarchy>
            <ExpanderHierarchy.Expander title="Nivå 1" onExpand={handleExpand}>
              {'hei'}
            </ExpanderHierarchy.Expander>
          </ExpanderHierarchy>
        );

        const button = screen.getByRole('button', { name: 'Nivå 1' });
        await userEvent.click(button);
        await userEvent.click(button);

        expect(handleExpand).toHaveBeenCalledTimes(2);
        expect(handleExpand).toHaveBeenNthCalledWith(1, true);
        expect(handleExpand).toHaveBeenNthCalledWith(2, false);
      });
    });

    describe('når ExpanderHierarchy har htmlMarkup=h3', () => {
      it('så har overskriften riktig nivå', () => {
        render(
          <ExpanderHierarchy htmlMarkup="h3">
            <ExpanderHierarchy.Expander title="Nivå 1">{'hei'}</ExpanderHierarchy.Expander>
          </ExpanderHierarchy>
        );

        const heading = screen.getByRole('heading', { name: 'Nivå 1', level: 3 });
        expect(heading).toBeVisible();
      });
    });

    describe('når Expander har expanded=true', () => {
      it('så vises innholdet i den fra starten av', async () => {
        render(
          <ExpanderHierarchy>
            <ExpanderHierarchy.Expander title="Nivå 1" expanded>
              {'hei'}
            </ExpanderHierarchy.Expander>
          </ExpanderHierarchy>
        );

        const heading = screen.getByRole('heading', { name: 'Nivå 1', level: 2 });
        expect(heading).toBeVisible();
        const button = screen.getByRole('button', { name: 'Nivå 1' });
        expect(button).toBeVisible();

        expect(button).toHaveAttribute('aria-expanded', 'true');

        const content = screen.getByText('hei');
        expect(content).toBeVisible();
      });
    });

    describe('når testId-props er satt', () => {
      it('så kan komponentene finnes med testId', async () => {
        render(
          <ExpanderHierarchy testId="list">
            <ExpanderHierarchy.Expander title="Nivå 1" testId="button">
              {'hei'}
            </ExpanderHierarchy.Expander>
          </ExpanderHierarchy>
        );

        const list = screen.getByTestId('list');
        expect(list).toBeVisible();
        const button = screen.getByTestId('button');
        expect(button).toBeVisible();
      });
    });
  });

  describe('gitt at ExpanderHierarchy har to nivåer', () => {
    describe('når man åpner alle nivåene', () => {
      it('så vises innholdet i dem', async () => {
        render(
          <ExpanderHierarchy>
            <ExpanderHierarchy.Expander title="Nivå 1">
              {'hei'}
              <ExpanderHierarchy>
                <ExpanderHierarchy.Expander title="Nivå 2">{'hei igjen'}</ExpanderHierarchy.Expander>
              </ExpanderHierarchy>
            </ExpanderHierarchy.Expander>
          </ExpanderHierarchy>
        );

        const heading = screen.getByRole('heading', { name: 'Nivå 1', level: 2 });
        expect(heading).toBeVisible();
        const button1 = screen.getByRole('button', { name: 'Nivå 1' });
        expect(button1).toBeVisible();

        expect(button1).toHaveAttribute('aria-expanded', 'false');

        await userEvent.click(button1);
        expect(button1).toHaveAttribute('aria-expanded', 'true');

        const content1 = screen.getByText('hei');
        expect(content1).toBeVisible();

        const button2 = screen.getByRole('button', { name: 'Nivå 2' });
        expect(button2).toBeVisible();

        expect(button2).toHaveAttribute('aria-expanded', 'false');

        await userEvent.click(button2);
        expect(button2).toHaveAttribute('aria-expanded', 'true');

        const content2 = screen.getByText('hei igjen');
        expect(content2).toBeVisible();
      });
    });
  });
});
