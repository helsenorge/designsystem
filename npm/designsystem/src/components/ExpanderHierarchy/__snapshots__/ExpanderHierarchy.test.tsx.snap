// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpanderHierarchy gitt at ExpanderHierarchy har en Expander når komponenten rendres med print=true så ser den riktig ut 1`] = `
<div>
  <ul
    class="expanderhierarchy expanderhierarchy--1"
    data-analyticsid="expander-hierarchy"
  >
    <li
      class="expander"
    >
      <h2
        class="expander__title expander__title--1 expander__title--print"
      >
        <button
          aria-expanded="false"
          class="expander__button expander__button--print"
          data-analyticsid="expander-hierarchy-expander"
          type="button"
        >
          Nivå 1
          <svg
            aria-hidden="true"
            class="hnds-style-icon expander__icon"
            data-analyticsid="icon"
            fill="black"
            focusable="false"
            height="38"
            role="presentation"
            style="min-width: 38px; min-height: 38px;"
            viewBox="0 0 48 48"
            width="38"
          >
            <path
              d="M32.577 17.885l1.793 1.779-10.371 10.451L13.63 19.664l1.793-1.779 8.576 8.644z"
            />
          </svg>
        </button>
      </h2>
      <div
        class="expander__content expander__content--1"
      >
        hei
      </div>
    </li>
  </ul>
</div>
`;

exports[`ExpanderHierarchy gitt at ExpanderHierarchy har en Expander når komponenten rendres så ser den riktig ut 1`] = `
<div>
  <ul
    class="expanderhierarchy expanderhierarchy--1"
    data-analyticsid="expander-hierarchy"
  >
    <li
      class="expander"
    >
      <h2
        class="expander__title expander__title--1"
      >
        <button
          aria-expanded="false"
          class="expander__button"
          data-analyticsid="expander-hierarchy-expander"
          type="button"
        >
          Nivå 1
          <svg
            aria-hidden="true"
            class="hnds-style-icon expander__icon"
            data-analyticsid="icon"
            fill="black"
            focusable="false"
            height="38"
            role="presentation"
            style="min-width: 38px; min-height: 38px;"
            viewBox="0 0 48 48"
            width="38"
          >
            <path
              d="M32.577 17.885l1.793 1.779-10.371 10.451L13.63 19.664l1.793-1.779 8.576 8.644z"
            />
          </svg>
        </button>
      </h2>
      <div
        class="expander__content expander__content--1"
      >
        hei
      </div>
    </li>
  </ul>
</div>
`;
